### ISSUE TITLE
inconsistent behavior around two-way binding and expressions

### ISSUE THREAD

**Issue by gefinity** [ractivejs/ractive#2030]
I'm using a function expression to get the value for a input field in a form based on some dynamic state like so:

`value = {{ getTranslation('name', currentLocale) }}`

which looks into something like: 

`name: {
    "en-GB": "some localised value",
    "en-US": "some other value"
}`

I knew two-way binding wouldn't work in this case so I'm using my own on-change/on-input event pair to update the state. 
However, I was expecting to see a single warning about expressions not working with two-way binding but instead get a completely different hard error, every time the you input data into the form.

`Computed properties without setters are read-only. (This may change in a future version of Ractive!)`

I messed around in fiddle and it seems if I use that expression directly on a input element I get what I expect but my input is actually inside a component.
See my fiddle:
http://jsfiddle.net/a7uafx0a/

Case 2 is more similar to my real setup while case 1 is a example of the warning I expected to see.

Why does ractive think the second case is a computed property?

I tried setting twoway=false and that would work but the problem is I can't seem to set twoway dynamically in any way. I've tried various expression forms and it only works if i hardcore true or false. I use this component in lots of places very usefully and sometimes it has a very straightforward value which can be bound and in some cases it isn't and I instead pass a 'onChange' callback to the component so I can deal with the sync manually.

If I could do this

`twoway="{{onChange? false : true}}"`

on the underlying input element inside the component I would be happy but... this doesn't work for me.

I could live with passing twoway=false on the components in this use case that doesn't seem to work either.

Thoughts?

**Comment by fskreuz** [comment]
So your main issue is trying to two-way derrived data. Instead of using a data function to get the translation, I suggest you do computed properties with an explicit getter and setter. That way, you get a chance to transform the data to display as well as setting it back.

```html
<input type="text" value="{{translatedText}}" />
<script>
component.exports = {
  computed: {
    translatedText: {
      get: function(){ return /* original -> translation */ },
      set: function(text) { this.set(/* translation -> original */); }
    }
  }
}
</script>
```

> Computed properties without setters are read-only. (This may change in a future version of Ractive!)

This happens when you use a computed property that doesn't provide a setter function on a two-way element (like inputs). By default, if you don't provide an explicit setter, all computed properties are getters (read-only).

Now the problem of this error message is it doesn't tell you which input element used the computed property *or where that computed property resides*. Yes, this can happen when you bind a read-only computed property of an ancestor component to a descendant component's input element.

Long story short: You might want to check the data source if it's a computed property from an ancestor and if it has a setter.

**Comment by gefinity** [comment]
My issue was really just confusion over why the 2 cases in the flddle produced different errors. A one time warning when ractive obviously couldn't twoway bind an expression and a spammy error. I didn't want ractive to attempt to two-way bind... computed properties with setters don't work for my case but I have another working solution. (apologies, the fiddle was over simplified to show the errors but in fact I need a parameter for my 'getter' so... it's just a normal function expression and I use a callback function on-change to respond to state change)
In any case I figured out how to control the twoway attr dynamically by setting it conditionally in my components onconfig method so the errors are silenced and everything works peacefully. Still not sure why the second case fails so differently, obviously something to do with using a component vs having the expression directly on an input so I thought it seemed confusing is all.

**Comment by evs-chris** [comment]
Since this is really old, and you can now change the `twoway` directive on the fly, I'm going to close this.
